libbpf-rs/src/program.rs

```diff
@@ -391,4 +399,11 @@ impl Program {
             Ok(Link::new(ptr))
         }
     }
+
+    /// Returns program's underlying BPF instructions.
+    ///
+    /// Please see note in [`Insns`].
+    pub fn insns(&self) -> Insns {
+        Insns::new(self.ptr)
```

Comment:
Interesting! Is there some reason we can't just return a Vec<Insn> to the user?

Reply:
That's what I am not sure about. It could return a Vec and then we could remove `Insns` and ignore the pointer to `bpf_insn` array but according to this [patch](https://lore.kernel.org/bpf/20211025224531.1088894-3-andrii@kernel.org/), the pointer array returned by `bpf_program__insns` can be used for [cloning](https://github.com/anakryiko/retsnoop/pull/1) so, to support this use-case, we'd need `Insns` to hold our pointer to the `bpf_insn` array. Alternatively, I guess we could add the ability to convert the Vec into a C array and use that for cloning?

Reply:
I pushed commit [b569bfe](https://github.com/libbpf/libbpf-rs/pull/203/commits/b569bfe847bfa6b9edede071428cd9876c46e155) which returns a Vec as you suggested. Let me know how that one looks. I'll squash the commits once we move forward.